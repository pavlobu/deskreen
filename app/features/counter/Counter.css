.backButton {
  position: absolute;
}

.counter {
  position: absolute !important;
  top: 30% !important;
  left: 45% !important;
  font-size: 10rem !important;
  font-weight: bold !important;
  letter-spacing: -0.025em !important;
}

.btnGroup {
  position: relative;
  top: 500px;
  width: 480px;
  margin: 0 auto;
}

.btn {
  font-size: 1.6rem;
  font-weight: bold;
  background-color: #fff;
  border-radius: 50%;
  margin: 10px;
  width: 100px;
  height: 100px;
  opacity: 0.7;
  cursor: pointer;
  font-family: Arial, Helvetica, Helvetica Neue, sans-serif;
}

.btn:hover {
  color: white;
  background-color: rgba(0, 0, 0, 0.5);
}
