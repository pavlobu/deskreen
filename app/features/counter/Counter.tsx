import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { Button } from '@blueprintjs/core';
import styles from './Counter.css';
import routes from '../../constants/routes.json';
import {
  increment,
  decrement,
  incrementIfOdd,
  incrementAsync,
  selectCount,
} from './counterSlice';

export default function Counter() {
  const dispatch = useDispatch();
  const value = useSelector(selectCount);
  return (
    <div>
      <div className={styles.backButton} data-tid="backButton">
        <br />
        <Link to={routes.HOME}>
          <Button icon="arrow-left" />
        </Link>
      </div>
      <h1
        className={`bp3-heading counter ${styles.counter}`}
        data-tid="counter"
      >
        {value}
      </h1>
      <div className={styles.btnGroup}>
        <Button
          className={styles.btn}
          onClick={() => {
            dispatch(increment());
          }}
          data-tclass="btn"
          type="button"
        >
          <i className="fa fa-plus" />
        </Button>
        <Button
          className={styles.btn}
          onClick={() => {
            dispatch(decrement());
          }}
          data-tclass="btn"
          type="button"
        >
          <i className="fa fa-minus" />
        </Button>
        <Button
          className={styles.btn}
          onClick={() => {
            dispatch(incrementIfOdd());
          }}
          data-tclass="btn"
          type="button"
        >
          odd
        </Button>
        <Button
          className={styles.btn}
          onClick={() => {
            dispatch(incrementAsync());
          }}
          data-tclass="btn"
          type="button"
        >
          async
        </Button>
      </div>
    </div>
  );
}
