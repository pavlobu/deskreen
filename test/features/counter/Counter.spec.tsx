/* eslint react/jsx-props-no-spreading: off, @typescript-eslint/ban-ts-comment: off */
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import Counter from '../../../app/features/counter/Counter';
import * as counterSlice from '../../../app/features/counter/counterSlice';

Enzyme.configure({ adapter: new Adapter() });
jest.useFakeTimers();

function setup(
  preloadedState: { counter: { value: number } } = { counter: { value: 1 } }
) {
  const store = configureStore({
    reducer: { counter: counterSlice.default },
    preloadedState,
  });

  const getWrapper = () =>
    mount(
      <Provider store={store}>
        <Router>
          <Counter />
        </Router>
      </Provider>
    );
  const component = getWrapper();
  return {
    store,
    component,
    buttons: component.find('button'),
    p: component.find('.counter'),
  };
}

describe('Counter component', () => {
  it('should should display count', () => {
    const { p } = setup();
    expect(p.text()).toMatch(/^1$/);
  });

  it('should 2 button should call increment', () => {
    const { buttons } = setup();
    const incrementSpy = jest.spyOn(counterSlice, 'increment');

    buttons.at(1).simulate('click');
    expect(incrementSpy).toBeCalled();
    incrementSpy.mockRestore();
  });

  it('should match exact snapshot', () => {
    const { store } = setup();
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Counter />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should 3 button should call decrement', () => {
    const { buttons } = setup();
    const decrementSyp = jest.spyOn(counterSlice, 'decrement');
    buttons.at(2).simulate('click');
    expect(decrementSyp).toBeCalled();
    decrementSyp.mockRestore();
  });

  it('should 3 button should call incrementIfOdd', () => {
    const { buttons } = setup();
    const incrementIfOdd = jest.spyOn(counterSlice, 'incrementIfOdd');
    buttons.at(3).simulate('click');
    expect(incrementIfOdd).toBeCalled();
    incrementIfOdd.mockRestore();
  });

  it('should 4 button should call incrementAsync', () => {
    const { buttons } = setup();
    const incrementAsync = jest.spyOn(counterSlice, 'incrementAsync');
    buttons.at(4).simulate('click');
    expect(incrementAsync).toBeCalled();
    incrementAsync.mockRestore();
  });

  it('should display updated count after increment button click', () => {
    const { buttons, p } = setup();
    buttons.at(0).simulate('click');
    expect(p.text()).toMatch(/^2$/);
  });

  it('should display updated count after decrement button click', () => {
    const { buttons, p } = setup();
    buttons.at(1).simulate('click');
    expect(p.text()).toMatch(/^0$/);
  });

  it('shouldnt change if even and if odd button clicked', () => {
    const { buttons, p } = setup({ counter: { value: 2 } });
    buttons.at(2).simulate('click');
    expect(p.text()).toMatch(/^2$/);
  });

  it('should change if odd and if odd button clicked', () => {
    const { buttons, p } = setup({ counter: { value: 1 } });
    buttons.at(2).simulate('click');
    expect(p.text()).toMatch(/^2$/);
  });
});

describe('Test counter actions', () => {
  it('should not call incrementAsync before timer', () => {
    const fn = counterSlice.incrementAsync(1000);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = jest.fn();
    // @ts-ignore
    fn(dispatch);
    jest.advanceTimersByTime(500);
    expect(dispatch).not.toBeCalled();
  });

  it('should call incrementAsync after timer', () => {
    const fn = counterSlice.incrementAsync(1000);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = jest.fn();
    // @ts-ignore
    fn(dispatch);
    jest.advanceTimersByTime(1001);
    expect(dispatch).toBeCalled();
  });
});
