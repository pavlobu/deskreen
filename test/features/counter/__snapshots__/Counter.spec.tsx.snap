// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component should match exact snapshot 1`] = `
<div
  className="container-fluid"
>
  <div
    className="backButton"
    data-tid="backButton"
  >
    <br />
    <a
      href="/"
      onClick={[Function]}
    >
      <button
        className="bp3-button"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        style={
          Object {
            "height": "50px",
            "width": "50px",
          }
        }
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-arrow-left"
          icon="arrow-left"
        >
          <svg
            data-icon="arrow-left"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              arrow-left
            </desc>
            <path
              d="M13.99 6.99H4.41L7.7 3.7a1.003 1.003 0 00-1.42-1.42l-5 5a1.014 1.014 0 000 1.42l5 5a1.003 1.003 0 001.42-1.42L4.41 8.99H14c.55 0 1-.45 1-1s-.46-1-1.01-1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </a>
  </div>
  <h1
    className="bp3-heading counter counter"
    data-tid="counter"
  >
    1
  </h1>
  <div
    className="btnGroup"
  >
    <button
      className="bp3-button btn"
      data-tclass="btn"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        <i
          className="fa fa-plus"
        />
      </span>
    </button>
    <button
      className="bp3-button btn"
      data-tclass="btn"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        <i
          className="fa fa-minus"
        />
      </span>
    </button>
    <button
      className="bp3-button btn"
      data-tclass="btn"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        odd
      </span>
    </button>
    <button
      className="bp3-button btn"
      data-tclass="btn"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        async
      </span>
    </button>
  </div>
</div>
`;
